modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/EagerTtlTest.java

```diff
@@ -0,0 +1,201 @@
+            .setAffinity(new RendezvousAffinityFunction(false, 16))
+            .setEagerTtl(eagerTtl)
+            .setWriteSynchronizationMode(CacheWriteSynchronizationMode.FULL_SYNC)
+            .setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL)
+            .setExpiryPolicyFactory(CreatedExpiryPolicy.factoryOf(new Duration(TimeUnit.SECONDS, 1)));
+    }
+
+    /**
+     * Checks of restart node with TTL cache.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testOneNodeRestartWithTtlCache() throws Exception {
+        eagerTtl = true;
+
+        IgniteEx ignite = startGrid(0);
+
+        ignite.cluster().state(ClusterState.ACTIVE);
+
+        IgniteCache cache = ignite.cache(DEFAULT_CACHE_NAME);
+
+        for (int i = 0; i < ENTRIES; i++)
+            cache.put(i, i);
+
+        ignite.close();
+
+        LogListener assertListener = LogListener.matches(ASSERTION_ERR).build();
+
+        listeningLog.registerListener(assertListener);
+
+        ignite = startGrid(0);
+
+        CountDownLatch exchnageHangLatch = new CountDownLatch(1);
+
+        ignite.context().cache().context().exchange().registerExchangeAwareComponent(new PartitionsExchangeAware() {
+            @Override public void onInitBeforeTopologyLock(GridDhtPartitionsExchangeFuture fut) {
+                try {
+                    exchnageHangLatch.await();
+                }
+                catch (InterruptedException e) {
+                    log.error("Interruped of waiting latch", e);
+
+                    fail(e.getMessage());
+                }
+            }
+        });
+
+        IgniteInternalFuture activeFut = GridTestUtils.runAsync(() -> ignite(0).cluster().state(ClusterState.ACTIVE));
+
+        assertFalse(activeFut.isDone());
+
+        assertFalse(GridTestUtils.waitForCondition(assertListener::check, 2_000));
+
+        exchnageHangLatch.countDown();
+
+        activeFut.get();
+
+        awaitPartitionMapExchange();
+    }
+
+    /**
+     * Checks that an assertion error does not happen during expiration in transactions.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testNotEagerEpireOnPut() throws Exception {
+        eagerTtl = false;
+
+        IgniteEx ignite = startGrid(0);
+
+        ignite.cluster().state(ClusterState.ACTIVE);
+
+        LogListener assertListener = LogListener.matches(ANY_ASSERTION_ERR).build();
+
+        listeningLog.registerListener(assertListener);
+
+        IgniteCache cache = ignite.cache(DEFAULT_CACHE_NAME);
+
+        for (int i = 0; i < ENTRIES; i++)
+            cache.put(i, i);
+
+        for (TransactionIsolation isolation : TransactionIsolation.values()) {
+            U.sleep(EXPIRATION_TIME);
+
+            for (int i = 0; i < ENTRIES; i++) {
+                try (Transaction tx = ignite.transactions().txStart(TransactionConcurrency.OPTIMISTIC, isolation)) {
+                    cache.putAll(Collections.singletonMap(i, isolation.ordinal() + 1));
```

Comment:
Could you please add `cache.getAndPut` as well?

Reply:
Done.